File: daytona-python-shell/repl.py
"""A tiny REPL that executes commands inside a Daytona sandbox.

This is a minimal REPL and is not a full shell — it has an incredibly limited subset of regular shell commands.
It is intended as a demonstration of how to interact with a Daytona sandbox programmatically.
"""

from __future__ import annotations

import posixpath
import sys

from daytona import Sandbox


def _expand_tilde(path: str, homedir: str) -> str:
    """Expand a leading ~ to the sandbox's home directory."""
    if path.startswith("~"):
        return posixpath.join(homedir, path[1:].lstrip("/"))
    return path


class ShellState:
    """Tracks the current working directory across commands."""

    def __init__(self, sandbox: Sandbox, homedir: str, cwd: str) -> None:
        self._sandbox = sandbox
        self._homedir = homedir
        self._cwd = cwd

    def run(self, cmd: str) -> tuple[int, str] | None:
        """Execute a shell command. Handles ``cd`` gracefully enough."""
        trimmed = cmd.strip()
        if not trimmed:
            return None

        parts = trimmed.split()
        if parts[0] == "cd":
            target = parts[1] if len(parts) > 1 else "~"
            target = _expand_tilde(target, self._homedir)
            if posixpath.isabs(target):
                self._cwd = target
            else:
                self._cwd = posixpath.join(self._cwd, target)
            return None

        result = self._sandbox.process.exec(f"cd {self._cwd} && {trimmed}")
        return (result.exit_code, result.result or "")


def tiny_daytona_repl(sandbox: Sandbox, *, cwd: str | None = None) -> None:
    """Spawn the REPL so that users can explore the sandbox interactively.

    Args:
        sandbox: The Daytona sandbox to execute commands in.
        cwd: The initial working directory inside the sandbox.
    """
    homedir = sandbox.process.exec("echo $HOME").result.strip()
    initial_cwd = _expand_tilde(cwd or "~", homedir)
    shell = ShellState(sandbox, homedir, initial_cwd)

    while True:
        try:
            line = input("$ ")
        except (EOFError, KeyboardInterrupt):
            # Ctrl+D or Ctrl+C — exit gracefully.
            print()
            break

        trimmed = line.strip()
        if not trimmed:
            continue
        if trimmed == "exit":
            break

        res = shell.run(trimmed)
        if res is None:
            continue

        exit_code, output = res

        if output:
            # Normalize the output to always end with a newline
            sys.stdout.write(output.removesuffix("\n") + "\n")

        if exit_code != 0:
            print(f"[exit {exit_code}]", file=sys.stderr)

File: blaxel-shell/repl.ts
import path from 'node:path';
import * as readline from 'node:readline';
import type { SandboxInstance } from '@blaxel/core';

/**
 * Options parameter for the @see tinyBlaxelRepl function.
 */
export interface IReplOptions {
  /** The current working directory. */
  cwd: string | undefined;
}

function expandTilde(filePath: string, homedir: string): string {
  return filePath.startsWith('~') ? path.posix.join(homedir, filePath.slice(1)) : filePath;
}

class ShellState {
  #homedir: string;
  #cwd: string;
  #sandbox: SandboxInstance;

  private constructor(homedir: string, cwd: string, sandbox: SandboxInstance) {
    this.#homedir = homedir;
    this.#cwd = cwd;
    this.#sandbox = sandbox;
  }

  static async create(sandbox: SandboxInstance, cwd: string): Promise<ShellState> {
    const homedir = await ShellState.probeHome(sandbox);
    return new ShellState(homedir, expandTilde(cwd, homedir), sandbox);
  }

  /** Execute a shell command. Handles `cd` gracefully enough. */
  async run(cmd: string): Promise<[number | null, string | null, string | null] | null> {
    const trimmed: string = cmd.trim();
    if (!trimmed) {
      return null;
    }

    const splitCommand = trimmed.split(/\s+/);
    if (splitCommand[0] === 'cd') {
      const target = splitCommand[1] ?? '~';
      this.#cwd = path.posix.isAbsolute(target)
        ? splitCommand[1]
        : path.posix.join(this.#cwd, this.expandTilde(target));
      return null;
    }

    const result = await this.#sandbox.process.exec({
      command: `cd ${this.#cwd} && ${trimmed}`,
      waitForCompletion: true,
    });
    return [result.exitCode, result.stdout, result.stderr];
  }

  private static async probeHome(sandbox: SandboxInstance): Promise<string> {
    const result = await sandbox.process.exec({ command: 'echo $HOME', waitForCompletion: true });
    return result.stdout.trim();
  }

  private expandTilde(filePath: string): string {
    return expandTilde(filePath, this.#homedir);
  }
}

function question(rl: readline.Interface, prompt: string): Promise<string | null> {
  return new Promise((resolve) => {
    rl.once('close', () => resolve(null));
    rl.question(prompt, (answer) => resolve(answer));
  });
}

/**
 * Spawns the REPL so that users can use it.
 *
 * This is a tiny REPL and is not a full shell, so it has an incredibly limited subset of regular shell commands.
 */
export default async function tinyBlaxelRepl(
  sandbox: SandboxInstance,
  options: IReplOptions | undefined
): Promise<void> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const shell = await ShellState.create(sandbox, options?.cwd ?? '~');

  while (true) {
    const input = await question(rl, '$ ');

    if (input === null) {
      break;
    }

    const res = await shell.run(input);
    if (res === null) {
      continue;
    }

    const [exitCode, stdout, stderr] = res;

    if (stdout) {
      process.stdout.write(stdout);
    }

    if (stderr) {
      process.stderr.write(stderr);
    }

    if (exitCode !== null && exitCode !== 0) {
      console.error(`The process exited with a non-zero exit code: ${exitCode}`);
    }
  }
}
